Assets/Mirror/Runtime/Batching/Batcher.cs

```diff
@@ -45,53 +51,75 @@ public Batcher(int threshold)
         // add a message for batching
         // we allow any sized messages.
         // caller needs to make sure they are within max packet size.
-        public void AddMessage(ArraySegment<byte> message)
+        public void AddMessage(ArraySegment<byte> message, double timeStamp)
         {
-            // put into a (pooled) writer
+            // when appending to a batch in progress, check final size.
+            // if it expands beyond threshold, then we should finalize it first.
+            // => less than or exactly threshold is fine.
+            //    GetBatch() will finalize it.
+            // => see unit tests.
+            if (batch != null &&
+                batch.Position + message.Count > threshold)
+            {
+                batches.Enqueue(batch);
+                batch = null;
+            }
+
+            // initialize a new batch if necessary
+            if (batch == null)
+            {
+                // borrow from pool. we return it in GetBatch.
+                batch = NetworkWriterPool.Get();
+
+                // write timestamp first.
+                // -> double precision for accuracy over long periods of time
+                batch.WriteDouble(timeStamp);
+            }
+
+            // add serialization to current batc. even if > threshold.
+            // -> we do allow > threshold sized messages as single batch
             // -> WriteBytes instead of WriteSegment because the latter
             //    would add a size header. we want to write directly.
-            // -> will be returned to pool when making the batch!
-            // IMPORTANT: NOT adding a size header / msg saves LOTS of bandwidth
-            NetworkWriterPooled writer = NetworkWriterPool.Get();
-            writer.WriteBytes(message.Array, message.Offset, message.Count);
-            messages.Enqueue(writer);
+            batch.WriteBytes(message.Array, message.Offset, message.Count);
         }
 
-        // batch as many messages as possible into writer
-        // returns true if any batch was made.
-        public bool MakeNextBatch(NetworkWriter writer, double timeStamp)
+        // helper function to copy a batch to writer and return it to pool
+        static void CopyAndReturn(NetworkWriterPooled batch, NetworkWriter writer)
```

Comment:
Why are we still making a copy?
While memory copies are fast, it would be faster to "do nothing" and just return/out param the pooled writer and have the caller return it to the pool

Reply:
yes, we can do that in a separate PR after.